File: examples/algorithms-python/src/leetcode/lc0647_palindromic_substrings.py
"""
LeetCode 647. Palindromic Substrings
链接: https://leetcode.com/problems/palindromic-substrings/
难度: Medium

题目描述:
给定一个字符串，计算这个字符串中有多少个回文子串。
具有不同开始位置或结束位置的子串，即使是由相同的字符组成，也会被视作不同的子串。

形式化规约:
  输入: s ∈ Σ*
  输出: 回文子串的数量

最优解思路:
  中心扩展法。每个回文子串都有一个中心（字符中心或间隙中心）。
  对于每个中心，向两边扩展，只要字符相等就计数。

复杂度:
  时间: O(n²)
  空间: O(1)

正确性要点:
  1. 2n-1 个中心覆盖所有奇偶长度的回文子串
  2. 每个中心向外扩展直到不匹配
  3. 每次扩展成功代表一个新的回文子串
"""


def count_substrings(s: str) -> int:
    """
    中心扩展法计数回文子串。
    时间复杂度 O(n²)，空间复杂度 O(1)。
    """
    n = len(s)
    count = 0

    for center in range(2 * n - 1):
        left = center // 2
        right = left + center % 2

        while left >= 0 and right < n and s[left] == s[right]:
            count += 1
            left -= 1
            right += 1

    return count


if __name__ == "__main__":
    # 示例 1: "a","b","c"
    assert count_substrings("abc") == 3, f"Example 1 failed: {count_substrings('abc')}"
    # 示例 2: "a","a","a","aa","aa","aaa"
    assert count_substrings("aaa") == 6, f"Example 2 failed: {count_substrings('aaa')}"
    # 边界: 空串
    assert count_substrings("") == 0, f"Empty failed: {count_substrings('')}"
    # 边界: 单字符
    assert count_substrings("a") == 1, f"Single char failed: {count_substrings('a')}"
    # 两个相同字符
    assert count_substrings("aa") == 3, f"Two same failed: {count_substrings('aa')}"
    # 两个不同字符
    assert count_substrings("ab") == 2, f"Two diff failed: {count_substrings('ab')}"
    # 偶回文
    assert count_substrings("abba") == 6, f"Even palindrome failed: {count_substrings('abba')}"

    print("All tests passed for LC 647 - Palindromic Substrings")

File: examples/algorithms/src/leetcode/lc0647_palindromic_substrings.rs
/// LeetCode 647. Palindromic Substrings
///
/// Given a string `s`, return the number of palindromic substrings in it.
/// A string is a palindrome when it reads the same backward as forward.
/// A substring is a contiguous sequence of characters within the string.
///
/// # Approach
/// Center expansion. For each possible center (2n-1 centers: n character centers
/// and n-1 gap centers between characters), expand outward while characters match.
/// Count each valid expansion as one palindromic substring.
///
/// # Correctness Proof
/// Every palindromic substring has a unique center (either a character for odd length,
/// or a gap for even length). The center expansion explores all possible centers,
/// and for each center, it expands as far as possible while the substring remains a palindrome.
/// Therefore, every palindromic substring is counted exactly once.
///
/// # Complexity
/// - Time complexity: O(n²)
/// - Space complexity: O(1)

pub struct Solution;

impl Solution {
    pub fn count_substrings(s: String) -> i32 {
        let bytes = s.as_bytes();
        let n = bytes.len();
        if n == 0 {
            return 0;
        }
        let mut count = 0;

        // Expand around each center
        for center in 0..(2 * n - 1) {
            let mut left = center / 2;
            let mut right = left + center % 2;

            while left < n && right < n && bytes[left] == bytes[right] {
                count += 1;
                if left == 0 {
                    break;
                }
                left -= 1;
                right += 1;
            }
        }

        count
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example_1() {
        assert_eq!(Solution::count_substrings("abc".to_string()), 3);
    }

    #[test]
    fn test_example_2() {
        assert_eq!(Solution::count_substrings("aaa".to_string()), 6);
    }

    #[test]
    fn test_empty() {
        assert_eq!(Solution::count_substrings("".to_string()), 0);
    }

    #[test]
    fn test_single_char() {
        assert_eq!(Solution::count_substrings("a".to_string()), 1);
    }

    #[test]
    fn test_two_same() {
        assert_eq!(Solution::count_substrings("aa".to_string()), 3); // "a", "a", "aa"
    }

    #[test]
    fn test_two_diff() {
        assert_eq!(Solution::count_substrings("ab".to_string()), 2); // "a", "b"
    }

    #[test]
    fn test_even_palindrome() {
        assert_eq!(Solution::count_substrings("abba".to_string()), 6); // "a","b","b","a","bb","abba"
    }
}
